docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,531 @@
+
+Non-networked endpoints (without an IP address) are considered local, and will only be advertised to other peers connecting via the same protocol. For example, an in-memory transport might use `Endpoint{Protocol: "memory", Path: "foo"}` as an address for the node "foo", and this should only be advertised to other nodes using `Protocol: "memory"`.
+
+#### Connections and Streams
+
+A connection represents an established transport connection between two endpoints (and thus two nodes), which can be used to exchange bytes via logically distinct IO streams. Connections are set up either via `Transport.Dial()` (outbound) or `Transport.Accept()` (inbound). The caller is responsible for verifying the remote peer's public key as returned by the connection.
+
+Data is exchanged over IO streams created with `Connection.Stream()` using an arbitrary `StreamID`. These implement the standard Go `io.Reader` and `io.Writer` interfaces to read and write bytes. Transports are free to choose how to implement such streams, e.g. by taking advantage of native stream support in the underlying protocol or through multiplexing.
+
+`Connection` and the related `Stream` interfaces are:
+
+```go
+// Connection represents an established connection between two endpoints.
+type Connection interface {
+    // Stream opens a logically distinct IO stream within the connection,
+    // using an arbitrary stream ID. The stream must be created before it
+    // can receive data, and must be closed before the ID can be reused.
+    Stream(StreamID) Stream
+
+    // LocalEndpoint returns the local endpoint for the connection.
+    LocalEndpoint() Endpoint
+
+    // RemoteEndpoint returns the remote endpoint for the connection.
+    RemoteEndpoint() Endpoint
+
+    // PubKey returns the public key of the remote peer.
+    PubKey() crypto.PubKey
+
+    // Close closes the connection.
+    Close() error
+}
+
+// StreamID is an arbitrary stream ID.
+type StreamID uint8
+
+// Stream represents a single logical IO stream within a connection.
+type Stream interface {
+    io.Reader // Read([]byte) (int, error)
+    io.Writer // Write([]byte) (int, error)
+    io.Closer // Close() error
+}
+```
+
+### Peers
+
+Peers are other Tendermint network nodes. Each peer is identified by a unique `PeerID`, and has a set of `PeerAddress` addresses expressed as URLs that they can be reached at. Examples of peer addresses might be e.g.:
+
+* `mconn://b10c@host.domain.com:25567/path`
+* `unix:///var/run/tendermint/peer.sock`
+* `memory:testpeer`
```

Comment:
Do Peers have one PeerAddress per Transport?

Reply:
No, they can have many. A peer has an arbitrary number of addresses, and each address uses an arbitrary transport (which may or may not be the same one).